### PR TITLE
Cache the pipeline

### LINKED ISSUE #74: Cache intermediate pipeline step results in the build pipeline
Should speed up the analysis process.

I think I'd like to tackle this by caching the result of each "step" in the pipeline within the docker container, so that if something does go wonky we can just restart the image to clear the caches.

- [ ] Rework the pipeline to be clearer about data sending / caching
- [ ] Automatically cache the output of each stage based on the (input?) content

### PR DESCRIPTION
**Pull request by hawkrives**
closes #74 

I don't even know what all to put here right now

This redesigns the entire pipeline structure around a declarative set of `{input, transform, output}` tuples that facilitate the automatic caching of each step's output

- You will get an array of inputs and you are expected to return an array of outputs
- The order of the `input` array is the order you will get your arguments
- The order of the `output` array is matched with the order of the returned array
- The colons serve the purpose of indicating what progress dot to fill out – everything after the colon is stripped while searching for a progress dot
- don't forget that you can specify multiple inputs and multiple outputs, and also that you can have two inputs and only generate one output
- if you need to send updated data back to the client (think the phylo graph) you can just add `:N` to the end of the step output key so you don't overwrite the prior info (in case something else wants it later)

Other things:

- The pipeline now tells the client how long each step took
- The client's websocket connection is now explicitly a global
- The newick textbox uses a second pipeline to process the newick data
- Results are cached on disk under the sha256 result of (the input/output key) + (the sha256 hash of the source input data)
- If a step used the cached data, its dot will turn green instead of blue

Todo:
- [ ] Fix the newick input box – it's broken

LMK with questions

### Example Pipeline

```js
const hybridizationSearch = [
	{
		// the first step: ensures that the input is converted to FASTA
		input: ['source'],
		transform: ([{ filepath, contents }]) =>
			filepath.endsWith('.fasta') ? [contents] : [genbankToFasta(contents)],
		output: ['initial-fasta'],
	},
	{
		// aligns the FASTA sequences
		input: ['initial-fasta'],
		transform: ([data]) => [clustal(data)],
		output: ['aligned-fasta'],
	},
	{
		// converts the aligned FASTA into Nexus
		input: ['aligned-fasta'],
		transform: ([data]) => [fastaToNexus(data)],
		output: ['aligned-nexus'],
	},
	{
		// does whatever mrbayes does
		input: ['aligned-nexus'],
		transform: ([data]) => [mrBayes(data)],
		output: ['consensus-tree'],
	},
	{
		// turns MrBayes' consensus tree into a Newick tree
		input: ['consensus-tree'],
		transform: ([data]) => [consensusTreeToNewick(data)],
		output: ['newick-tree'],
	},
	{
		// turns the Newick tree into a JSON object
		input: ['newick-tree'],
		transform: ([data]) => [parseNewick(data)],
		output: ['newick-json:1'],
	},
	{
		input: ['newick-json:1', 'aligned-fasta'],
		transform: ([newickJson, alignedFasta]) => {
			let { removedData, prunedNewick } = pruneOutliers(
				newickJson,
				alignedFasta
			)
			return [prunedNewick, removedData]
		},
		output: ['newick-json:2', 'pruned-identifiers'],
	},
	{
		// identifies the non-monophyletic sequences
		input: ['newick-json:2', 'aligned-fasta'],
		transform: ([newickJson, alignedFasta]) => [
			removeCircularLinks(ent.strictSearch(newickJson, alignedFasta)),
			removeCircularLinks(newickJson),
		],
		output: ['nonmonophyletic-sequences', 'newick-json:3'],
	},
]
```

## Before

hybsearch with all dots green due to not caching

## After

hybsearch with all dots green due to caching

### CODE REVIEW

**Comment by hawkrives** [comment]
Oh, one other thing is that the results of every step are now sent back to the client. The client doesn't need to do something with all of them, but it just seemed simpler to send it all back.

File: server/pipeline/cache.js [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+		this.set = this.set.bind(this)
+		this._dataHash = this._hash(this.data)
+
+		this.set('source', { filepath: this.filepath, contents: this.data })
+	}
+
+	_hash(data) {
+		let hash = crypto.createHash('sha256')
+		hash.update(data)
+		return hash.digest('hex')
```
Comment (rye) [inline server/pipeline/cache.js]: I suppose since it's a digest it doesn't matter, but I'm partial to `base64`—no aversion to this, though.

Comment (hawkrives) [reply server/pipeline/cache.js]: I don't even know what options I can give it. The docs only ever showed "hex".

**Comment by hawkrives** [comment]
I think I want to make one tweak before we merge – I'd like to add the contents of the `transform` function to the hash key computation, so that if you change how the output is computed you'll invalidate the cache.

**Comment by hawkrives** [comment]
Okay, I've realized that I don't actually care about hashing the `transform` function, for one reason: even if we hash it, these functions are barely wrappers around other functions that actually do the work, and we can't (easily) hash the contents of those functions. And those are the bits more likely to change.

So I'd rather let this be a simpler implementation than cover a minor case while leaving the major one unhandled.

**Pull request merged**